fleetspeak/src/client/daemonservice/execution/execution.go

```diff
@@ -453,3 +473,40 @@ func (e *Execution) statsLoop() {
 	// This blocks until the daemon process terminates.
 	rum.Run()
 }
+
+// heartbeatMonitorRoutine monitors the daemon process's hearbeats and kills
+// unresponsive processes.
+func (e *Execution) heartbeatMonitorRoutine(pid int) {
+	defer e.inProcess.Done()
+
+	// Give the child process some time to start up. During boot it sometimes
+	// takes significantly more time than the unresponsive_kill_period to start
+	// the child so we disable checking for heartbeats for a while.
+	e.recordHeartbeat()
+	sleepTime := e.heartbeatUnresponsiveGracePeriod
+
+	for {
```

Comment:
My reading of this loop is that it keeps going until a heartbeat fails, even if the execution is otherwise over. This is wasteful, and with a long heartbeat period problematic - besides being a hanging goroutine, other parts of the system may be waiting for us to call inProcess.Done().

I think the right fix is to change every sleep to a select which waits for either a timer to fire (in which case we continue) or e.Done to be closed (in which case we return).

Reply:
Hmmm.. thinkin about it more waiting for e.dead instead of e.Done might be slightly better, but this is a fine point and either would get the job done.

Reply:
Done.